File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/AddProxyPresenter.java
package dev.ragnarok.fenrir.mvp.presenter;

import static dev.ragnarok.fenrir.util.Utils.trimmedIsEmpty;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import dev.ragnarok.fenrir.Injection;
import dev.ragnarok.fenrir.mvp.presenter.base.RxSupportPresenter;
import dev.ragnarok.fenrir.mvp.view.IAddProxyView;
import dev.ragnarok.fenrir.settings.IProxySettings;
import dev.ragnarok.fenrir.util.ValidationUtil;


public class AddProxyPresenter extends RxSupportPresenter<IAddProxyView> {

    private final IProxySettings settings;
    private boolean authEnabled;
    private String address;
    private String port;
    private String userName;
    private String pass;

    public AddProxyPresenter(@Nullable Bundle savedInstanceState) {
        super(savedInstanceState);
        settings = Injection.provideProxySettings();
    }

    @Override
    public void onGuiCreated(@NonNull IAddProxyView viewHost) {
        super.onGuiCreated(viewHost);
        viewHost.setAuthFieldsEnabled(authEnabled);
        viewHost.setAuthChecked(authEnabled);
    }

    public void fireAddressEdit(CharSequence s) {
        address = s.toString();
    }

    public void firePortEdit(CharSequence s) {
        port = s.toString();
    }

    public void fireAuthChecked(boolean isChecked) {
        if (authEnabled == isChecked) {
            return;
        }

        authEnabled = isChecked;
        callView(v -> v.setAuthFieldsEnabled(isChecked));
    }

    public void fireUsernameEdit(CharSequence s) {
        userName = s.toString();
    }

    public void firePassEdit(CharSequence s) {
        pass = s.toString();
    }

    private boolean validateData() {
        try {
            try {
                int portInt = Integer.parseInt(port);
                if (portInt <= 0) {
                    throw new Exception("Invalid port");
                }
            } catch (NumberFormatException e) {
                throw new Exception("Invalid port");
            }

            if (!ValidationUtil.isValidIpAddress(address) && !ValidationUtil.isValidURL(address)) {
                throw new Exception("Invalid address");
            }

            if (authEnabled && trimmedIsEmpty(userName)) {
                throw new Exception("Invalid username");
            }

            if (authEnabled && trimmedIsEmpty(pass)) {
                throw new Exception("Invalid password");
            }
        } catch (Exception e) {
            callView(v -> showError(v, e));
            return false;
        }

        return true;
    }

    public void fireSaveClick() {
        if (!validateData()) {
            return;
        }

        String finalAddress = address.trim();
        int finalPort = Integer.parseInt(port.trim());

        if (authEnabled) {
            settings.put(finalAddress, finalPort, userName, pass);
        } else {
            settings.put(finalAddress, finalPort);
        }

        callView(IAddProxyView::goBack);
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/addproxy/AddProxyPresenter.kt
package dev.ragnarok.fenrir.fragment.addproxy

import android.os.Bundle
import dev.ragnarok.fenrir.Includes.proxySettings
import dev.ragnarok.fenrir.fragment.base.RxSupportPresenter
import dev.ragnarok.fenrir.settings.IProxySettings
import dev.ragnarok.fenrir.trimmedIsNullOrEmpty
import dev.ragnarok.fenrir.util.ValidationUtil.isValidIpAddress
import dev.ragnarok.fenrir.util.ValidationUtil.isValidURL

class AddProxyPresenter(savedInstanceState: Bundle?) :
    RxSupportPresenter<IAddProxyView>(savedInstanceState) {
    private val settings: IProxySettings = proxySettings
    private var authEnabled = false
    private var address: String? = null
    private var port: String? = null
    private var userName: String? = null
    private var pass: String? = null
    override fun onGuiCreated(viewHost: IAddProxyView) {
        super.onGuiCreated(viewHost)
        viewHost.setAuthFieldsEnabled(authEnabled)
        viewHost.setAuthChecked(authEnabled)
    }

    fun fireAddressEdit(s: CharSequence?) {
        address = s.toString()
    }

    fun firePortEdit(s: CharSequence?) {
        port = s.toString()
    }

    fun fireAuthChecked(isChecked: Boolean) {
        if (authEnabled == isChecked) {
            return
        }
        authEnabled = isChecked
        view?.setAuthFieldsEnabled(isChecked)
    }

    fun fireUsernameEdit(s: CharSequence?) {
        userName = s.toString()
    }

    fun firePassEdit(s: CharSequence?) {
        pass = s.toString()
    }

    private fun validateData(): Boolean {
        try {
            try {
                val portInt = port?.toInt() ?: 0
                if (portInt <= 0) {
                    throw Exception("Invalid port")
                }
            } catch (e: NumberFormatException) {
                throw Exception("Invalid port")
            }
            if (!isValidIpAddress(address) && !isValidURL(address)) {
                throw Exception("Invalid address")
            }
            if (authEnabled && userName.trimmedIsNullOrEmpty()) {
                throw Exception("Invalid username")
            }
            if (authEnabled && pass.trimmedIsNullOrEmpty()) {
                throw Exception("Invalid password")
            }
        } catch (e: Exception) {
            showError(e)
            return false
        }
        return true
    }

    fun fireSaveClick() {
        if (!validateData()) {
            return
        }
        val finalAddress = address?.trim { it <= ' ' }
        val finalPort = port?.trim { it <= ' ' }?.toInt() ?: 0
        if (authEnabled) {
            if (finalAddress != null) {
                userName?.let {
                    pass?.let { it1 ->
                        settings.put(
                            finalAddress,
                            finalPort,
                            it,
                            it1
                        )
                    }
                }
            }
        } else {
            if (finalAddress != null) {
                settings.put(finalAddress, finalPort)
            }
        }
        view?.goBack()
    }

}